### ISSUE TITLE
batched cholesky decomposition GPU vs. CPU speed

### ISSUE THREAD

**Issue by matthiasreisser** [pytorch/pytorch#20700]
## 🐛 Bug

Batched Cholesky Decomposition is much slower on GPU than on CPU. For small number of batches, CPU is orders of magnitude faster than the corresponding GPU implementation
## To Reproduce
```
def cholesky_speed_test(device="cpu"):
    assert device in ("cpu","cuda")
    np.random.seed(123)
    n_rep = 10000
    dim = 16
    batches = 10

    A = np.random.normal(0,1,(batches,dim,dim))
    cov = np.matmul(A,A.swapaxes(1,2))

    pt_cov = torch.tensor(cov.astype(np.float32),device=device)
    L = torch.cholesky(pt_cov)

    torch.cuda.nvtx.range_push("CholeskyDecomposition")
    start_time = time.time()
    for i in range(n_rep):
        torch.cuda.nvtx.range_push("Iter{}".format(i))
        torch.cholesky(pt_cov, out=L)
        torch.cuda.nvtx.range_pop()
    duration = time.time() - start_time
    print("Duration: {:.3f}s on device {}".format(duration,device))
    torch.cuda.nvtx.range_pop()
```

1. Execute the above function for device "cpu" and "cuda"
2. We observe the following times on a 1080Ti
```
Duration: 5.007s on device cuda
Duration: 0.253s on device cpu
```
3. Investigating Nvidia Nsight logs:
![image](https://user-images.githubusercontent.com/8858505/58018090-2f95ac80-7b02-11e9-9d2f-dcf236588716.png)
Shows that `torch.cholesky` is allocating and freeing memory constantly.

## Expected behavior

I would expect the GPU code to be much faster and not spend so much time on memory allocation

## Environment

Collecting environment information... 
PyTorch version: 1.1.0 
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 16.04.4 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.9) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 9.0.176
GPU models and configuration: GPU 0: GeForce GTX 1080 Ti
Nvidia driver version: 384.130
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.0.5

Versions of relevant libraries:
[pip3] numpy==1.15.4
[pip3] torch==1.1.0
[pip3] torchvision==0.2.2
[conda] blas                      1.0                         mkl
[conda] mkl                       2019.1                      144
[conda] mkl_fft                   1.0.10           py36ha843d7b_0
[conda] mkl_random                1.0.2            py36hd81dba3_0
[conda] pytorch                   1.1.0           py3.6_cuda9.0.176_cudnn7.5.1_0    pytorch
[conda] torchvision               0.2.2                      py_3    pytorch

Any help or insight is appreciated!

Matthias

cc @vincentqb @vishwakftw @jianyuh @nikitaved @pearu @mruberry @heitorschueroff @VitalyFedyunin @ngimel

**Comment by vishwakftw** [comment]
Hi Matthias,

Thank you for raising this issue.

I've noticed that using GPU for larger batches and/or matrices is more effective. For instance, I ran the function that you have provided with n_rep = 100, batches = 100 and dim = 160 , and the GPU implementation outperformed the CPU implementation by over a second (cumulatively).

**Comment by benelot** [comment]
I am currently looking at the performance of the different solvers, among them the cholesky solver. I use it together with other operations to perform calculations on batches. Is there any reason why the solver uses more and more of the total CPU and GPU time the larger the batch?


For a batch of 8:
```
--------------------------------------------------------------------------------
  autograd profiler output (CUDA mode)
--------------------------------------------------------------------------------
        top 15 events sorted by cpu_time_total

	Because the autograd profiler uses the CUDA event API,
	the CUDA time column reports approximately max(cuda_time, cpu_time).
	Please ignore this output if your code does not use CUDA.

------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------
Name                Self CPU total %   Self CPU total      CPU total %        CPU total     CPU time avg     CUDA total %       CUDA total    CUDA time avg  Number of Calls
------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------
cholesky_solve               38.26%        938.782us           38.26%        938.782us        938.782us           37.20%        944.064us        944.064us                1
cholesky                     33.50%        821.863us           33.50%        821.863us        821.863us           32.58%        826.912us        826.912us                1
mul                           2.74%         67.345us            2.74%         67.345us         67.345us            2.73%         69.344us         69.344us                1
einsum                        2.68%         65.767us            2.68%         65.767us         65.767us            2.80%         71.040us         71.040us                1
diag_embed                    2.66%         65.267us            2.66%         65.267us         65.267us            3.03%         76.960us         76.960us                1
to                            2.56%         62.812us            2.56%         62.812us         62.812us            2.43%         61.728us         61.728us                1
cat                           2.48%         60.866us            2.48%         60.866us         60.866us            2.61%         66.336us         66.336us                1
cat                           2.39%         58.585us            2.39%         58.585us         58.585us            2.41%         61.088us         61.088us                1
cat                           2.12%         52.089us            2.12%         52.089us         52.089us            2.30%         58.432us         58.432us                1
matmul                        2.00%         49.060us            2.00%         49.060us         49.060us            2.82%         71.521us         71.521us                1
matmul                        1.88%         46.053us            1.88%         46.053us         46.053us            1.78%         45.280us         45.280us                1
mul                           1.77%         43.495us            1.77%         43.495us         43.495us            1.25%         31.648us         31.648us                1
cat                           1.69%         41.464us            1.69%         41.464us         41.464us            1.87%         47.456us         47.456us                1
cat                           1.66%         40.667us            1.66%         40.667us         40.667us            2.44%         61.888us         61.888us                1
cat                           1.61%         39.380us            1.61%         39.380us         39.380us            1.74%         44.224us         44.224us                1
------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------
Self CPU time total: 2.453ms
CUDA time total: 2.538ms

```

Batch of 10 * 1024 (to exagerate the effect):
```
--------------------------------------------------------------------------------
  autograd profiler output (CUDA mode)
--------------------------------------------------------------------------------
        top 15 events sorted by cpu_time_total

	Because the autograd profiler uses the CUDA event API,
	the CUDA time column reports approximately max(cuda_time, cpu_time).
	Please ignore this output if your code does not use CUDA.

------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------
Name                Self CPU total %   Self CPU total      CPU total %        CPU total     CPU time avg     CUDA total %       CUDA total    CUDA time avg  Number of Calls
------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------
cholesky_solve               81.47%         32.224ms           81.47%         32.224ms         32.224ms           90.70%         31.848ms         31.848ms                1
cholesky                     16.58%          6.557ms           16.58%          6.557ms          6.557ms            6.14%          2.156ms          2.156ms                1
to                            0.31%        123.819us            0.31%        123.819us        123.819us            0.35%        122.400us        122.400us                1
einsum                        0.23%         91.104us            0.23%         91.104us         91.104us            0.54%        190.464us        190.464us                1
to                            0.20%         80.325us            0.20%         80.325us         80.325us            0.23%         80.352us         80.352us                1
diag_embed                    0.15%         59.589us            0.15%         59.589us         59.589us            0.49%        170.400us        170.400us                1
mean                          0.13%         52.665us            0.13%         52.665us         52.665us            0.17%         60.800us         60.800us                1
rsub                          0.13%         50.322us            0.13%         50.322us         50.322us            0.16%         56.704us         56.704us                1
scalar_tensor                 0.12%         48.217us            0.12%         48.217us         48.217us            0.06%         21.152us         21.152us                1
sigmoid                       0.12%         46.671us            0.12%         46.671us         46.671us            0.18%         63.136us         63.136us                1
add                           0.12%         45.759us            0.12%         45.759us         45.759us            0.13%         45.184us         45.184us                1
mul                           0.11%         44.727us            0.11%         44.727us         44.727us            0.16%         57.664us         57.664us                1
mean                          0.11%         43.765us            0.11%         43.765us         43.765us            0.37%        130.176us        130.176us                1
cat                           0.11%         43.402us            0.11%         43.402us         43.402us            0.17%         60.508us         60.508us                1
cat                           0.10%         40.504us            0.10%         40.504us         40.504us            0.15%         51.968us         51.968us                1
------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------
Self CPU time total: 39.553ms
CUDA time total: 35.116ms

```

It seems that the solver does not scale as well with the batch size as the other ops. Why is that? One would believe that each batch can be solved in parallel with separate linear system solver instances and thus it would scale well.

**Comment by eusoubrasileiro** [comment]
@matthiasreisser I faced the same behavior but I think that it is expected.

As you pointed in (I saw exactly the same)
> 3. Shows that torch.cholesky is allocating and freeing memory constantly.

I read somewhere that GPU memory allocation and deallocation **is slower** raw compared to CPU memory manipulations, altough I don't have the reference here right now.

**Comment by ngimel** [comment]
Closing, for 10x16x16 you can't expect improvements from cuda, for bigger batch sizes cuda is significantly faster.
